feat(backups): validate VHD streams

julien-f opened the pull request:
### Check list

> Check if done, if not relevant leave unchecked.

- [ ] PR reference the relevant issue (e.g. `Fixes #007` or `See xoa-support#42`)
- [ ] if UI changes, a screenshot has been added to the PR
- [ ] documentation updated
- `CHANGELOG.unreleased.md`:
  - [ ] enhancement/bug fix entry added
  - [ ] list of packages to release updated (`${name} v${new version}`)
- **I have tested added/updated features** (and impacted code)
  - [ ] unit tests (e.g. [`cron/parse.spec.js`](https://github.com/vatesfr/xen-orchestra/blob/b24400b21de1ebafa1099c56bac1de5c988d9202/%40xen-orchestra/cron/src/parse.spec.js))
  - [ ] if `xo-server` API changes, the corresponding test has been added to/updated on [`xo-server-test`](https://github.com/vatesfr/xen-orchestra/tree/master/packages/xo-server-test)
  - [x] at least manual testing

### Process

1. create a PR as soon as possible
1. mark it as `WiP:` (Work in Progress) if not ready to be merged
1. when you want a review, add a reviewer (and only one)
1. if necessary, update your PR, and re- add a reviewer

From [_the Four Agreements_](https://en.wikipedia.org/wiki/Don_Miguel_Ruiz#The_Four_Agreements):

1. Be impeccable with your word.
1. Don't take anything personally.
1. Don't make assumptions.
1. Always do your best.

julien-f committed:
WiP: vhdStreamValidator

fbeauchamp commented on @xen-orchestra/backups/_VmBackup.js:
```diff
@@ -251,6 +255,11 @@ class VmBackup {
     const sizeContainers = mapValues(deltaExport.streams, stream => watchStreamSize(stream))
+
+    if (this._settings.validateVhdStreams) {
+      deltaExport.streams = mapValues(deltaExport.streams, stream => pipeline(stream, vhdStreamValidator, noop))
```
I think it may remove the `_nbd` property used to show in the UI that nbd is used . Maybe we should put this property in the VM object ?

(previous changes like throttle may already have the same effect)

julien-f replied:
If that's the case, it may have already been broken by the speed throttling.

I think it's not a good practice to add properties to streams and we should probably find better ways to do this.

julien-f replied:
If I understand correctly, the NBD test is before, therefore this change does not break it.

fbeauchamp commented on packages/vhd-lib/_StreamReader.js:
```diff
@@ -0,0 +1,127 @@
+    stream.on('readable', () => this.#resolve())
+
+    finished(stream, { writable: false }, error_ => {
+      this.#error = error_
```
```suggestion
    finished(stream, { writable: false }, error => {
      this.#error = error
```

fbeauchamp commented on packages/vhd-lib/_StreamReader.js:
```diff
@@ -0,0 +1,127 @@
+
+      // per Node documentation:
+      // > The size argument must be less than or equal to 1 GiB.
+      assert(size < 1073741824)
```
Why are we testing it if it is already enforced by the stream implementation ?

fbeauchamp commented on packages/vhd-lib/vhdStreamValidator.js:
```diff
@@ -0,0 +1,50 @@
+    last = chunk
+  }
+
+  assert.deepEqual(last, footer1)
```
I think we should decorate the error with the number of bytes read (to detect if errors occurs on the same place) , and the number of second since the start and last read  (to detect timeout )

fbeauchamp reviewed (changes requested):
I think we should remove the validation done on `RemoteAdapter.outputStream`, since it will be the same check

julien-f committed:
fixes

julien-f committed:
better diff

- save only expected data and diff
- save expected buffer as hex string for compacity and readability
- save diff as an array of differences represented as hex strings

julien-f committed:
split into packages

julien-f committed:
put sizes in strict error messages

julien-f committed:
test stream errors

julien-f committed:
better var name

julien-f committed:
add usages
